cv32e40s/tests/programs/custom/debug_test_trigger/debug_test_trigger.c

```diff
@@ -0,0 +1,1738 @@
+  __asm__ volatile ("csrwi scontext, 0x0");
+  if (!illegal_insn_status) return FAIL;
+
+  execute_debug_command(DEBUG_SEL_ENTER_MACHINEMODE);
+
+  return SUCCESS;
+}
+
+/*
+ * pmp_setup
+ *
+ * PMP setup function
+ *
+ * Allows access to full memory map from user mode
+ */
+void pmp_setup(void) {
+  // Allow user mode access to full memory map
+  __asm__ volatile (R"(li    t0, 0xFFFFFFFF
+                       csrw  pmpaddr0, t0
+                       csrwi pmpcfg0, ((1 << 3) | (7 << 0))
+                       )" ::: "t0");
+}
+
+/*
+ * get_num_triggers
+ *
+ * Get number of triggers
+ *
+ * Determine number of triggers implemented by probing tselect
+ *
+ */
+uint32_t get_num_triggers() {
+  illegal_insn_status = 0;
+  __asm__ volatile ("csrwi tselect, 0x0");
+
+  if (illegal_insn_status) {
+    num_triggers = 0;
+  } else {
+    __asm__ volatile (R"(
+      csrwi tselect, 0x1
+      csrwi tselect, 0x2
+      csrwi tselect, 0x3
+      csrr s2, tselect
+      la   s3, num_triggers
+      sw   s2, 0(s3)
+
+      csrwi tselect, 0x0
+    )" ::: "s2", "s3");
+
+    num_triggers++;
+  }
+
+  printf ("NUM_TRIGGERS = %ld\n", num_triggers);
+
+  return num_triggers;
+}
+
+/*
+ * main
+ *
+ * Test Entry point
+ *
+ */
+int main(int argc, char *argv[])
+{
+  pmp_setup();
+  num_triggers = get_num_triggers();
+
+  if (num_triggers > 0) {
+    for (int i = 0; i < num_triggers; i++) {
+
+      trigger_sel = i;
+      printf ("csr_write: tselect = %ld\n", trigger_sel);
+      __asm__ volatile (R"(lw        s2, trigger_sel
+                           csrw tselect, s2         )" ::: "s2");
+
+      if (test_register_access()) {
+        printf("Register access test failed\n");
+        return FAIL;
+      }
+      if (test_execute_trigger(PRIV_LVL_MACHINE_MODE)) {
+        printf("Execute trigger test (machine mode) failed\n");
+        return FAIL;
+      }
+      if (test_execute_trigger(PRIV_LVL_USER_MODE)) {
+        printf("Execute trigger test (user mode) failed\n");
+        return FAIL;
+      }
+      if (test_load_trigger(PRIV_LVL_MACHINE_MODE)) {
+        printf("Load trigger test (machine mode) failed\n");
+        return FAIL;
+      }
+      if (test_load_trigger(PRIV_LVL_USER_MODE)) {
+        printf("Load trigger (user mode) test failed\n");
+        return FAIL;
+      }
+      if (test_store_trigger(PRIV_LVL_MACHINE_MODE)) {
+        printf("Store trigger test (machine mode) failed\n");
+        return FAIL;
+      }
+      if (test_store_trigger(PRIV_LVL_USER_MODE)) {
+        printf("Store trigger (user mode) test failed\n");
+        return FAIL;
+      }
+      if (test_exception_trigger(PRIV_LVL_MACHINE_MODE)) {
+        printf("Exception trigger test (machine mode) failed\n");
+        return FAIL;
+      }
+      if (test_exception_trigger(PRIV_LVL_USER_MODE)) {
+        printf("Exception trigger (user mode) test failed\n");
+        return FAIL;
+      }
+    }
+    printf("Finished \n");
+    return SUCCESS;
+  } else {
```

Comment:
If num_triggers==0 only returns FAIL, where is cfg0 used?

Reply:
> Checks for DBG_NUM_TRIGGERS==0 are separated out in another test and will be committed in separate PR.
It is used with the debug_test_0_triggers test which will be PRed when this is PR is merged.
It was just easier to add all the configs at the same time.